src/nf/nf_multihead_attention_submodule.f90

```diff
@@ -0,0 +1,343 @@
+    call self % query_layer % backward(self % q_input, self % combine_heads(dq))
+
+    ! free temporary storages
+    deallocate(d_output)
+    deallocate(v_heads)
+    deallocate(k_heads)
+    deallocate(q_heads)
+    deallocate(d_sdpa)
+    deallocate(jacobian)
+    deallocate(d_normalize)
+    deallocate(dq)
+    deallocate(dk)
+  end subroutine common_backward
+
+  pure module subroutine common_forward(self, query, key, value)
+    class(multihead_attention_layer), intent(in out) :: self
+    real, intent(in) :: query(:, :), key(:, :), value(:, :)
+
+    real, allocatable :: q(:, :, :)
+    real, allocatable :: k(:, :, :)
+    real, allocatable :: v(:, :, :)
+
+    ! allocate storage for intermidiate stages
+    allocate(q(self % sequence_length, self % head_size, self % n_heads))
+    allocate(k(self % sequence_length, self % head_size, self % n_heads))
+    allocate(v(self % sequence_length, self % head_size, self % n_heads))
+
+    self % q_input = query
+    self % k_input = key
+    self % v_input = value
+
+    ! run inputs through linear layers (trainable params)
+    call self % query_layer % forward(query)
+    call self % key_layer % forward(key)
+    call self % value_layer % forward(value)
+
+    ! split attention heads for more efficient computation
+    q = self % split_heads(self % query_layer % output)
+    k = self % split_heads(self % key_layer % output)
+    v = self % split_heads(self % value_layer % output)
+
+    ! create key by value matrix
+    call self % create_attention_matrix(q, k)
+    ! apply softmax and scaling
+    call self % normalize_attention_matrix()
+    ! multiply attention matrix by value
+    call self % scaled_dot_product_attention(v)
+
+    self % o_input = self % combine_heads(self % sdpa)
+    call self % output_layer % forward(self % o_input)
+    self % output = self % output_layer % output
+
+    ! free temp vars from memory
+    deallocate(q)
+    deallocate(k)
+    deallocate(v)
+  end subroutine common_forward
+
+  pure module function split_heads(self, input) result(output)
+    class(multihead_attention_layer), intent(in) :: self
+    real, intent(in) :: input(:, :)
+    real :: output(self % sequence_length, self % head_size, self % n_heads)
+    output = reshape(input, [self % sequence_length, self % head_size, self % n_heads])
+  end function split_heads
+
+  pure module subroutine create_attention_matrix(self, query, key)
+    class(multihead_attention_layer), intent(in out) :: self
+    real, intent(in) :: query(:, :, :)
+    real, intent(in) :: key(:, :, :)
+    integer :: head
+    ! create attention matrix for each sequence in each batch
+    do concurrent(head = 1: self % n_heads)
+      self % attention_matrix(:, :, head) = matmul(query(:, :, head), transpose(key(:, :, head)))
+    end do
+  end subroutine create_attention_matrix
+
+  pure module subroutine normalize_attention_matrix(self, attention_mask)
```

Comment:
`attention_mask` is not accessible to the users by design at this point. It will be used by transformer decoder later and I'll add corresponding logic later